### ISSUE TITLE
Installing on ARM64 shows an error that the AMD64 version of the DLL couldn't be registered

### ISSUE THREAD

**Issue by kristibektashi** [GerbilSoft/rom-properties#398]
The AMD64 version of the DLL doesn't have to be registered at all if the ARM64EC version exists (and they might even conflict). 
![Screenshot 2023-12-03 080819](https://github.com/GerbilSoft/rom-properties/assets/40217426/5a0b5eed-0073-4e25-b8c8-b10bee8a4760)

**Comment by GerbilSoft** [comment]
I don't have any ARM Windows machines so I can't test this myself, but I made an adjustment to the install program such that if Windows 11 is detected, it registers i386, arm64, and arm64ec instead of i386, amd64, and arm64.

Please try install.exe from this build once it completes: https://ci.appveyor.com/project/GerbilSoft/rom-properties/builds/48666841/job/wvlgruu7afne0ufx
(Copy it into the v2.2.1 release directory, since the AppVeyor build only has a single architecture.)

(Note that install.exe is always i386 for maximum compatibility, since all versions of Windows 10 on ARM support i386 emulation.)

**Comment by kristibektashi** [comment]
> I don't have any ARM Windows machines so I can't test this myself, but I made an adjustment to the install program such that if Windows 11 is detected, it registers i386, arm64, and arm64ec instead of i386, amd64, and arm64.
> 
> Please try install.exe from this build once it completes: https://ci.appveyor.com/project/GerbilSoft/rom-properties/builds/48666841/job/wvlgruu7afne0ufx (Copy it into the v2.2.1 release directory, since the AppVeyor build only has a single architecture.)
> 
> (Note that install.exe is always i386 for maximum compatibility, since all versions of Windows 10 on ARM support i386 emulation.)

https://github.com/GerbilSoft/rom-properties/commit/92393c3a20fcb2145bcf1a10088fbc1ed8e2343e#commitcomment-134084953 (I made this comment before I saw your reply)

**Comment by GerbilSoft** [comment]
Checking again, it seems amd64 emulation was never actually part of Windows 10; it was only released for Windows 11. I'll rework that and post a new build momentarily.

https://blogs.windows.com/windows-insider/2020/12/10/introducing-x64-emulation-in-preview-for-windows-10-on-arm-pcs-to-the-windows-insider-program/

**Comment by GerbilSoft** [comment]
With regards to ARM32 deprecation in build 25905: This page says ARM32 UWP applications are deprecated. It doesn't say anything specifically about ARM32 Win32 applications.

https://blogs.windows.com/windows-insider/2023/07/12/announcing-windows-11-insider-preview-build-25905/

**Comment by GerbilSoft** [comment]
New build will be here: https://ci.appveyor.com/project/GerbilSoft/rom-properties/builds/48666950/job/4traenmdo9lai3ru

Note that it may take a while to actually build.

**Comment by kristibektashi** [comment]
> With regards to ARM32 deprecation in build 25905: This page says ARM32 UWP applications are deprecated. It doesn't say anything specifically about ARM32 Win32 applications.
> 
> https://blogs.windows.com/windows-insider/2023/07/12/announcing-windows-11-insider-preview-build-25905/

ARM32 Win32 apps were never officially supported by Microsoft, but nevertheless they can be compiled with Visual Studio and will still work fine on builds before 25905. And it has been confirmed that both UWP and Win32 app support has been removed for ARM32 in build 25905 despite the official changelog not mentioning it (I believe the sysarm32 folder doesn't exist anymore)

Edit: https://betawiki.net/wiki/Windows_11_build_25905#Miscellaneous

**Comment by GerbilSoft** [comment]
Updated: https://ci.appveyor.com/project/GerbilSoft/rom-properties/builds/48667998/job/8e83718i9wbswp1t

The 32-bit ARM DLL will no longer be registered on Win11 build 25905 and later.

...and now that I'm looking at it more closely, I'm pretty sure that code path is only hit if running on a 32-bit ARM build of Windows 10, so it's essentially a no-op. Probably should have it register the 32-bit ARM version on 64-bit ARM Windows 10, but I guess only if I receive a bug report, since I don't think many 32-bit Windows ARM applications are available in the first place.

**Comment by GerbilSoft** [comment]
One more update: https://ci.appveyor.com/project/GerbilSoft/rom-properties/builds/48668072/job/w3524k0rwn0s00au

The previous commit didn't actually make sense because if IsWow64Process2() returns 32-bit ARM, that means it's running on a 32-bit ARM system, which isn't possible on Windows 11. (I believe Windows 10 does support this.)

Reworked the check for the CPU_arm64 path, along with adding 32-bit ARM registration if running on Win11 earlier than build 25905.

**Comment by GerbilSoft** [comment]
v2.7 has a proper installer using Inno Setup. On ARM64 Windows systems, it will install the i386, arm, arm64, and arm64ec DLLs; the amd64 DLLs won't be installed or registered. (...though it seems I did this a while back too; the Inno Setup version should be more resilient.)

**Comment by kristibektashi** [comment]
> v2.7 has a proper installer using Inno Setup. On ARM64 Windows systems, it will install the i386, arm, arm64, and arm64ec DLLs; the amd64 DLLs won't be installed or registered. (...though it seems I did this a while back too; the Inno Setup version should be more resilient.)

You should look into building the ARM64 DLL as ARM64X, which is a binary format that includes both ARM64 and ARM64EC code in the same file, and it also covers x64 code as well since ARM64EC does that anyway. That way you will only have to register the i386, ARM(32) and ARM64X DLL's but it will still cover everything.

Also note that afaik the ARM64 and ARM64EC/x64 DLL's cannot be registered at the same time, since unlike i386 or ARM32, x64 apps on ARM64 don't use WOW64 and share the same registry space as ARM64 applications. ARM64X is the workaround for this, since the same DLL can be loaded on ARM64/ARM64EC/x64 processes all at once.

Here's the Microsoft doc on ARM64X: https://learn.microsoft.com/en-us/windows/arm/arm64x-pe

**Comment by GerbilSoft** [comment]
Checking the documents, it looks like ARM64X is backwards-compatible with Windows 10 on ARM (which doesn't otherwise support ARM64EC). So it looks like I should simply drop arm64/arm64ec and replace it with arm64x. (I wasn't aware of the registration limitation, so that means previous versions were definitely doing something wrong.)

**Comment by kristibektashi** [comment]
> Checking the documents, it looks like ARM64X is backwards-compatible with Windows 10 on ARM (which doesn't otherwise support ARM64EC). So it looks like I should simply drop arm64/arm64ec and replace it with arm64x. (I wasn't aware of the registration limitation, so that means previous versions were definitely doing something wrong.)

Yes but you still need to target ARM64 and ARM64EC in your project so that the ARM64X DLL can be built correctly.

**Comment by GerbilSoft** [comment]
Currently modifying the packaging script to build ARM64 and ARM64EC with the modifications needed for ARM64X.
Reference: https://learn.microsoft.com/en-us/windows/arm/arm64x-build

Code changes:
* The ARM64EC portion of the DLLs will now check for the arm64/ and arm64ec/ subdirectories, instead of just arm64ec/ .

package.cmd changes:
* Build ARM64 with the ARM64X option enabled. This doesn't change the ARM64 build by itself, but it makes the information available to the ARM64EC build.
* Build ARM64EC with the ARM64X option enabled. This builds ARM64X DLLs with ARM64 as the default architecture.

setup6.iss changes:
* Install DLLs from the ARM64EC build and EXEs from the ARM64 build. (We don't need ARM64EC executables. ARM64EC executables are only needed in cases where AMD64 DLLs need to be loaded into the process.)
* All of these are installed in the arm64/ directory, instead of having both arm64/ and arm64ec/ .

**Comment by GerbilSoft** [comment]
@kristibektashi Can you test this installer? It's built from commit 26b59cf0c6b88d6b9f016da77da6058efb4e4aba.

[SetupRomProperties.master-26b59cf0.zip](https://github.com/user-attachments/files/22788023/SetupRomProperties.master-26b59cf0.zip)

It's an all-arch installer so it should work on i386, amd64, and arm64; not arm32 because Windows for 32-bit ARM doesn't support i386 applications. (Also, it's using Inno Setup 6, so it requires Windows 7 or later. Not an issue for ARM testing, obviously...)

On an arm64 system, it should install the following in C:\Program Files\rom-properties:
- i386/
- arm/ (on Windows 10; this won't be installed on Windows 11)
- arm64/ (with ARM64X DLLs)
- ARM64-architecture EXEs

You can use rom-properties to check its own files after installation. (The ARM64X DLLs should show up as "CHPEv2 ARM64X".)

**Comment by kristibektashi** [comment]
@GerbilSoft ARM64 versions of Windows 11 up to build 25398 (unleaked) should still support ARM32 applications, so I think the ARM32 binaries should be installed on those versions of Windows as well

**Comment by kristibektashi** [comment]
Also sorry but I can't test right now, my ARM64 device (Raspberry Pi 4B) is giving me trouble, and I can't be bothered to fix it right now

**Comment by GerbilSoft** [comment]
New test build: [SetupRomProperties-v2.6+.master-f75b0c7a.zip](https://github.com/user-attachments/files/22867974/SetupRomProperties-v2.6%2B.master-f75b0c7a.zip)

This build needs testing on arm and arm64 systems. I managed to compile the GNU gettext runtime library for arm and arm64x (arm64+arm64ec), and it *seems* to be working (everything linked correctly with MSVC 17.14.16), but since I don't have an ARM Windows system, I can't test it.

**Comment by GerbilSoft** [comment]
Closing as fixed for now. ARM64X is broken in v2.7 and will be temporarily reverted in v2.7.1; will try fixing it later. (Plain ARM64 will work properly in v2.7.1.)
